### ISSUE TITLE
U-net 2D (multi-label) training issue

### ISSUE THREAD

**Issue by eddin022** [HenriquesLab/ZeroCostDL4Mic#155]
First, I want to say how impressive the notebooks and wikipage are. With that out of the way, I'm having a bit of an issue with training in the U-net 2D (multilabel) notebook. I want to point out that the steps prior to training have all run successfully for me.

The training loss decreases with an increase in epoch number and plateaus as expected, however the validation loss remains "nan" throughout the training session. This might be an easy fix or an error on my part, but I can't seem to figure out a solution. 

At the end of the training session I receive a weights_last.hdf5 file, but no weights_best.hdf5 file (presumably because of the 'nan' validation loss problem.

Is there anything I can try to fix the problem? - Thanks in advance!

Casey

**Comment by esgomezm** [comment]
Hi Casey, 

Thank you very much for giving it a try. 

There migth be different reasons why this is happening. The most common one is when there's only background in your validation images. For that, could you please share a screenshot of the displays you get while the network is training (cell 4.2).

Yet another reason is when there're not enough images to create the validation dataset. How many images do you have in the training set and what was the percentage you use for the validation split?

Sincerely,
Esti

**Comment by eddin022** [comment]
Thanks for such a speedy response!
Attached are images 1) first epoch 2) last epoch
Training set # is 197 images - I have tried both a 10% and 20% validation split.

Casey

![Screen Shot 2021-11-22 at 9 25 32 AM](https://user-images.githubusercontent.com/67563125/142888674-cd9319ef-549d-46ff-8abd-718da4a5dd27.png)
![Screen Shot 2021-11-22 at 9 24 52 AM](https://user-images.githubusercontent.com/67563125/142888691-a8de3e02-b649-4d3e-8e66-5413b813a9bf.png)

**Comment by eddin022** [comment]
Ahh! I think I solved it and I think it was the format of my data. I played with the formatting and got it to work!

Thanks so much for the help,
Casey

**Comment by esgomezm** [comment]
Hi Casey, 

so it's getting complicated... The only reason I can come with is that the validation data chosen by the method is empty (belonging to background) or that there are images with pixel values different from 0, 1 and 2.

Is this happening every time you run the notebook? Could you please run all the cells again and check if it happens all the time? Just seeing 2 epochs should be enough. Also, could you try to reduce the learning rate drammatically (e.g., 0.000001) to see if at least in the first epochs it calculates some loss value? 

Cheers, 
Esti

**Comment by eddin022** [comment]
Dear Esti,

I think the problem was the mask pixel values were 1,3,5 instead of 0,1,2.
I played with the format of my data and the problem is solved!

Thanks for your help,
Casey

**Comment by esgomezm** [comment]
Great! Thank you for giving it a try ;)

**Comment by UthpalaK** [comment]
Hello, I have the same issue. My error is always nan. I have tried to reduce the learning rate, but the same thing happens. I have seen the issue's author mentioned that he changed the mask pixel values. I would like to know how to change it or how to check my mask pixel values. Thank you in advance

![image](https://github.com/HenriquesLab/ZeroCostDL4Mic/assets/11831589/94d21d76-e927-42c6-99be-4d1eed2cbdf6)
![image](https://github.com/HenriquesLab/ZeroCostDL4Mic/assets/11831589/63f6ac57-89e4-49ee-93f5-2814a45deb81)

**Comment by esgomezm** [comment]
Dear @UthpalaK 
thank you for giving it a try. Could you please check the format of your training masks? The image should be 8 or 16 bits, but the pixel values should be 0 for the background, 1 for the little structures (gray) and 2 for the big structures (white). Could you maybe open one of the images in fiji and see if those values are like that? Also, how many labels did you include in the notebook?

**Comment by esgomezm** [comment]
@UthpalaK I'm going to close this issue, but please, if you still find issues with the notebook, do not hesitate opening it again. Thank you!
